src/eval/datetime.rs

```diff
@@ -0,0 +1,175 @@
+use std::fmt;
+use std::fmt::{Debug, Formatter};
+use std::hash::Hash;
+
+use ecow::{eco_format, eco_vec, EcoString, EcoVec};
+use time::error::{Format, InvalidFormatDescription};
+use time::format_description;
+
+use crate::util::pretty_array_like;
+use typst_macros::cast_from_value;
+
+/// A datetime object that represents either a date, a time or a combination of
+/// both.
+#[derive(Clone, Copy, PartialEq, Hash)]
+pub enum Datetime {
+    /// Representation as a date.
+    Date(time::Date),
+    /// Representation as a time.
+    Datetime(time::PrimitiveDateTime),
+    /// Representation as a combination of date and time.
+    Time(time::Time),
+}
+
+impl Datetime {
+    /// Display the date and/or time in a certain format.
+    pub fn display(&self, pattern: Option<EcoString>) -> Result<EcoString, EcoString> {
+        let pattern = pattern.unwrap_or(match self {
+            Datetime::Date(_) => EcoString::from("[year]-[month]-[day]"),
+            Datetime::Time(_) => EcoString::from("[hour]:[minute]:[second]"),
+            Datetime::Datetime(_) => {
+                EcoString::from("[year]-[month]-[day] [hour]:[minute]:[second]")
+            }
+        });
+
+        let format = format_description::parse(pattern.as_str())
```

Comment:
Would the change below work? (To avoid constructing and then unwrapping EcoString unnecessarily)

```suggestion
        let pattern = pattern.map(EcoString::as_str).unwrap_or(match self {
            Datetime::Date(_) => "[year]-[month]-[day]",
            Datetime::Time(_) => "[hour]:[minute]:[second]",
            Datetime::Datetime(_) => {
                "[year]-[month]-[day] [hour]:[minute]:[second]"
            }
        });

        let format = format_description::parse(pattern)
```